src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1325,33 +1322,69 @@ object Desugar {
+              for {
+                l <- permissionD(ctx)(left.asInstanceOf[PExpression])
+                r <- permissionD(ctx)(right.asInstanceOf[PExpression])
+              } yield in.UneqCmp(l, r)(src)
+            } else {
+              for {
+                l <- exprAndTypeAsExpr(ctx)(left)
+                r <- exprAndTypeAsExpr(ctx)(right)
+              } yield in.UneqCmp(l, r)(src)
+            }
+
+          case PLess(left, right) =>
+            if (info.typ(left) == PermissionT || info.typ(right) == PermissionT) {
+              for {l <- permissionD(ctx)(left); r <- permissionD(ctx)(right)} yield in.PermLtCmp(l, r)(src)
+            } else {
+              for {l <- go(left); r <- go(right)} yield in.LessCmp(l, r)(src)
+            }
+
+          case PAtMost(left, right) =>
+            if (info.typ(left) == PermissionT || info.typ(right) == PermissionT) {
+              for {l <- permissionD(ctx)(left); r <- permissionD(ctx)(right)} yield in.PermLeCmp(l, r)(src)
+            } else {
+              for {l <- go(left); r <- go(right)} yield in.AtMostCmp(l, r)(src)
+            }
 
-          case PLess(left, right) => for {l <- go(left); r <- go(right)} yield in.LessCmp(l, r)(src)
-          case PAtMost(left, right) => for {l <- go(left); r <- go(right)} yield in.AtMostCmp(l, r)(src)
-          case PGreater(left, right) => for {l <- go(left); r <- go(right)} yield in.GreaterCmp(l, r)(src)
-          case PAtLeast(left, right) => for {l <- go(left); r <- go(right)} yield in.AtLeastCmp(l, r)(src)
+          case PGreater(left, right) =>
+            if (info.typ(left) == PermissionT || info.typ(right) == PermissionT) {
+              for {l <- permissionD(ctx)(left); r <- permissionD(ctx)(right)} yield in.PermGtCmp(l, r)(src)
+            } else {
+              for {l <- go(left); r <- go(right)} yield in.GreaterCmp(l, r)(src)
+            }
+
+          case PAtLeast(left, right) =>
+            if (info.typ(left) == PermissionT || info.typ(right) == PermissionT) {
+              for {l <- permissionD(ctx)(left); r <- permissionD(ctx)(right)} yield in.PermGeCmp(l, r)(src)
+            } else {
+              for {l <- go(left); r <- go(right)} yield in.AtLeastCmp(l, r)(src)
+            }
 
           case PAnd(left, right) => for {l <- go(left); r <- go(right)} yield in.And(l, r)(src)
           case POr(left, right) => for {l <- go(left); r <- go(right)} yield in.Or(l, r)(src)
 
           case PAdd(left, right) =>
             for {
-              l <- go(left);
+              l <- go(left)
               r <- go(right)
-              res = if (info.typ(left) == StringT && info.typ(right) == StringT) {
-                in.Concat(l, r)(src)
-              } else {
-                in.Add(l, r)(src)
+              res = (info.typ(left), info.typ(right)) match {
+                case (StringT, StringT) => in.Concat(l, r)(src)
+                case _ => in.Add(l, r)(src)
```

Comment:
`left` or `right` cannot be of type `PermissionT` as these cases are handled (with an early return) by `maybePermissionD`, can they?

Reply:
If either `left` or `right` are of type `PermissionT`, the type-checker/inference assigns the type `PermissionT` to the `PAdd` expression (assuming that the expression is well-typed). As such, those cases will be handled in the early return statement, before the match block.